### PR TITLE
Issue #648 prj file multiplier

### LINKED ISSUE #648: PRJ-parser does not use multiplication factor on grid input
In GitLab by @HendrikKok on Nov 16, 2023, 16:04

At least for STO-package, the multiplication factor is not used on gridinput. 

Input like:

1,2, 037,  0.1000000E-04,   0.000000    ,  -999.9900    ,'wvp19_thick.idf'

results in a storage coefficient that’s equal to the thickness of the aquifer.

Requirements
---------------
- Multiplication and addition should be able to be used by all packages
- Multiplication and addition should be possible per layer

Refinement
-------------
- Add test with multiplication and addition factors in projectfile
- Add functinality to multiply or add factors to constant/grids
- Addition and multiplication factors should be parsed in the existing project file reader, in ``imod.formats.prj``

Possible solution
-------------------
We could collect the addition factors, as well as multiplication factors, in a xr.DataArray with just a layer dimension. Then xarray's broadcasting can be used.

```python
...
addition_factors = xr.DataArray(addition_values, coords={"layer": layers}, dims = ("layer",)})
data = addition_factors + data
...
```

Example data
---------------
Example data is provided in the zip as collected in ``imod.data.imod5_projectfile_data``.
[Downloaded from artifacts page](https://github.com/Deltares/imod-artifacts/raw/main/). In the zip is .prj file which can be modified to test if the added functionality works.

### PR DESCRIPTION
**Pull request by luitjansl**
Fixes #648 

# Description
The project file which we use to import mf5 models into imod-python can contain multiplication factors and an addition
constants. These are provided for each layer separately (each layer is imported from a separate file, eg an idf file). When loading data, these  multipliers and addition constants should be applied to te data in the idf (or other type) file.

# Checklist

- [X] Links to correct issue
- [X] Update changelog, if changes affect users
- [X] PR title starts with ``Issue #nr``, e.g. ``Issue #737``
- [X] Unit tests were added
- [ ] **If feature added**: Added/extended example

### CODE REVIEW

File: imod/tests/test_mf6/test_import_prj.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    path=pooch.os_cache("imod"),
+    base_url="https://github.com/Deltares/imod-artifacts/raw/main/",
+    version=None,
+    version_dev="main",
+    env="IMOD_DATA_DIR",
+)
+
+with importlib.resources.files("imod.data") as pkg_dir:
+    REGISTRY.load_registry(pkg_dir / "registry.txt")
+fname_model = REGISTRY.fetch("iMOD5_model.zip")
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_import_prj.py:20]: This code is duplicate with the code in imod.data. I think there is more opportunity for code sharing.

File: imod/tests/test_mf6/test_import_prj.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    )
+
+
+def snippet_idf_import_kh(factor: float, addition: float):
+    return f"""
+0001,(KHV),1, Horizontal Permeability
+001,2
+1,1,1,{factor},{addition},-999.99, '.\Database\KHV\VERSION_1\IPEST_KHV_L1.IDF' >>> (KHV) Horizontal Permeability (IDF) <<<
+1,1,2,{factor},{addition},-999.99, '.\Database\KHV\VERSION_1\IPEST_KHV_L2.IDF' >>> (KHV) Horizontal Permeability (IDF) <<<
+"""
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_import_prj.py:67]: See comment about ``textwrap.dedent``

File: imod/tests/test_mf6/test_import_prj.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    )
+
+
+def snippet_gen_import_hfb(factor: float, addition: float):
+    return f"""\
+0001,(HFB),1, Horizontal Flow Barrier
+001,2
+ 1,2, 003,{factor},{addition},  -999.9900    ,'.\Database\HFB\VERSION_1\IBV2_HOOFDBREUKEN_BX.GEN' >>> (HFB) Horizontal Barrier Flow (GEN) <<<
+ 1,2, 005,{factor},{addition},  -999.9900    ,'.\Database\HFB\VERSION_1\IBV2_HOOFDBREUKEN_SY.GEN' >>> (HFB) Horizontal Barrier Flow (GEN) <<<
+"""
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_import_prj.py:105]: See comment about textwrap.dedent

File: imod/tests/test_mf6/test_import_prj.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    factor1: float, addition1: float, factor2: float, addition2: float
+):
+    return f"""\
+0001,(WEL),1, Wells
+STEADY-STATE
+001,003
+1,2,5,{factor1},{addition1},-999.99,                                       '.\Database\WEL\VERSION_1\WELLS_L3.IPF' >>> (WRA) Well Rate (IPF) <<<
+1,2,7,{factor2},{addition2},-999.99,                                       '.\Database\WEL\VERSION_1\WELLS_L4.IPF' >>> (WRA) Well Rate (IPF) <<<
+1,2,9,{factor1},{addition1},-999.99,                                       '.\Database\WEL\VERSION_1\WELLS_L5.IPF' >>> (WRA) Well Rate (IPF) <<<
+"""
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_import_prj.py:133]: See comment about textwrap.dedent

**Review by JoerivanEngelen — COMMENTED** [review]
Looks good, I have some comments on the tests

File: imod/tests/test_mf6/test_import_prj.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+fname_model = REGISTRY.fetch("iMOD5_model.zip")
+
+
+def snippet_constant_kh(factor: float, addition: float, init: float):
+    return f"""
+0001,(KHV),1, Horizontal Permeability
+001,2
+1,1,1,{factor},{addition},{init}   >>> (KHV) Horizontal Permeability (IDF) <<<
+1,1,2,{factor},{addition},{init}   >>> (KHV) Horizontal Permeability (IDF) <<<
+"""
```
[the remark this answers is not in the document]
Comment (luitjansl) [reply imod/tests/test_mf6/test_import_prj.py]: done

File: imod/tests/test_mf6/test_import_prj.py (hunk 2 above)
Comment (luitjansl) [reply imod/tests/test_mf6/test_import_prj.py]: done

File: imod/tests/test_mf6/test_import_prj.py (hunk 3 above)
Comment (luitjansl) [reply imod/tests/test_mf6/test_import_prj.py]: done

File: imod/tests/test_mf6/test_import_prj.py (hunk 4 above)
Comment (luitjansl) [reply imod/tests/test_mf6/test_import_prj.py]: done

File: imod/tests/test_mf6/test_import_prj.py (hunk 1 above)
Comment (luitjansl) [reply imod/tests/test_mf6/test_import_prj.py]: done

File: imod/data/sample_data.py [hunk 6]
```diff
@@ -22,50 +22,61 @@
...
+
+
+def load_pooch_registry(registry: pooch.core.Pooch) -> pooch.core.Pooch:
+    with importlib.resources.files("imod.data") as pkg_dir:
+        registry.load_registry(pkg_dir / "registry.txt")
+    return registry
+
+
+registry = create_pooch_registry()
+registry = load_pooch_registry(registry)
```
Comment (JoerivanEngelen) [inline imod/data/sample_data.py:44]: I think it is better to make registry all caps here, to denote it as a global constant, as it is going to be used in the following functions, without it being an argument.

```suggestion
REGISTRY = create_pooch_registry()
REGISTRY = load_pooch_registry(REGISTRY)
```

File: imod/formats/prj/prj.py [hunk 7]
```diff
@@ -556,6 +556,24 @@ def _create_dataarray(
...
+        da = da * factor
+        da = da + addition
+    else:
+        nlayers = len(da.coords["layer"])
+        for ilayer in range(nlayers):
+            layer = da.coords["layer"][ilayer].values[()]
+            factor = headers[ilayer]["factor"]
+            addition = headers[ilayer]["addition"]
+            da.loc[{"layer": layer}] = da.sel(layer=layer).values * factor
+            da.loc[{"layer": layer}] = da.sel(layer=layer).values + addition
```
Comment (JoerivanEngelen) [inline imod/formats/prj/prj.py:576]: I think this works better with dask, as it prevents loading data in memory to apply factors and additions:

```suggestion
    layer, addition_values = list(zip([(i+1, header["addition"]) for i, header in enumerate(headers)]))
    factor_values = [header["factor"] for header in headers]
    addition = xr.DataArray(addition_values, coords = {"layer" : layer}, dims=("layer",))
    factor = xr.DataArray(factor_values, coords = {"layer" : layer}, dims=("layer",)
    da = da * factor + addition
```

**Review by JoerivanEngelen — COMMENTED** [review]
2 more comments

File: imod/data/sample_data.py (hunk 6 above)
Comment (luitjansl) [reply imod/data/sample_data.py:44]: done

**Pull request merged**